File: Python/dirchk.py
import os
import shutil
from collections import defaultdict

def format_size(size):
    """Convert bytes into the largest appropriate unit."""
    for unit in ['B', 'KB', 'MB', 'GB', 'TB', 'PB']:
        if size < 1024:
            return f"{size:.2f} {unit}"
        size /= 1024
    return f"{size:.2f} PB"

def get_disk_usage(path):
    """Recursively calculate disk usage for each directory and file."""
    total_size = 0
    ext_usage = defaultdict(int)

    try:
        for entry in os.scandir(path):
            if entry.is_dir(follow_symlinks=False):
                # Recursively get size for directories
                sub_size, sub_ext_usage = get_disk_usage(entry.path)
                total_size += sub_size
                for ext, size in sub_ext_usage.items():
                    ext_usage[ext] += size
            else:
                # Get size for files
                size = entry.stat().st_size
                total_size += size

                # Get file extension
                _, ext = os.path.splitext(entry.name)
                ext_usage[ext] += size
    except PermissionError:
        print(f"Permission Denied: {path}")

    return total_size, ext_usage

def print_tree_view(path, depth=0, total_space=0):
    """Print the directory tree view with percentage usage."""
    size, _ = get_disk_usage(path)
    percentage = (size / total_space) * 100 if total_space else 0
    print(f"{'  ' * depth}{os.path.basename(path)}/ - {format_size(size)} ({percentage:.2f}%)")

    try:
        for entry in os.scandir(path):
            if entry.is_dir(follow_symlinks=False):
                print_tree_view(entry.path, depth + 1, total_space)
    except PermissionError:
        print(f"{'  ' * (depth + 1)}Permission Denied: {path}")

def print_sorted_extensions(ext_usage):
    """Print a sorted list of file extensions by total usage."""
    sorted_ext = sorted(ext_usage.items(), key=lambda x: x[1], reverse=True)
    print("\nFile Extension Usage:")
    for ext, size in sorted_ext:
        print(f"{ext}: {format_size(size)}")

def print_disk_info(path):
    """Print the total, used, and free space of the disk where the path resides."""
    total, used, free = shutil.disk_usage(path)
    print(f"\nDisk Overview:")
    print(f"Total space: {format_size(total)}")
    print(f"Used space:  {format_size(used)}")
    print(f"Free space:  {format_size(free)}\n")

def main():
    path = input("Enter the directory path to analyze: ")

    # Print disk space info for the path
    print_disk_info(path)

    total_size, ext_usage = get_disk_usage(path)

    print("\nDisk Usage Tree View:")
    print_tree_view(path, total_space=total_size)

    print_sorted_extensions(ext_usage)

if __name__ == "__main__":
    main()

File: C++/dirchk.cpp
#include <iostream>
#include <filesystem>
#include <map>
#include <unordered_map>
#include <iomanip> // for setprecision

namespace fs = std::filesystem;

std::string format_size(uintmax_t size) {
    const char* units[] = {"B", "KB", "MB", "GB", "TB", "PB"};
    int unit_index = 0;
    double converted_size = size;

    while (converted_size >= 1024 && unit_index < 5) {
        converted_size /= 1024;
        ++unit_index;
    }
    
    std::ostringstream out;
    out << std::fixed << std::setprecision(2) << converted_size << " " << units[unit_index];
    return out.str();
}

struct DiskUsage {
    uintmax_t size = 0;
    std::unordered_map<std::string, uintmax_t> ext_usage;
};

DiskUsage get_disk_usage(const fs::path& path) {
    DiskUsage du;
    try {
        for (const auto& entry : fs::directory_iterator(path)) {
            if (entry.is_directory()) {
                // Recursively calculate directory size
                DiskUsage sub_dir_usage = get_disk_usage(entry.path());
                du.size += sub_dir_usage.size;
                for (const auto& [ext, size] : sub_dir_usage.ext_usage) {
                    du.ext_usage[ext] += size;
                }
            } else if (entry.is_regular_file()) {
                // Calculate file size
                uintmax_t size = entry.file_size();
                du.size += size;
                
                // Get file extension
                std::string ext = entry.path().extension().string();
                du.ext_usage[ext] += size;
            }
        }
    } catch (const fs::filesystem_error& e) {
        std::cerr << "Error accessing: " << path << " (" << e.what() << ")\n";
    }
    return du;
}

void print_tree_view(const fs::path& path, int depth, uintmax_t total_size) {
    DiskUsage du = get_disk_usage(path);
    double percentage = total_size ? (static_cast<double>(du.size) / total_size * 100) : 0;
    std::cout << std::string(depth * 2, ' ') << path.filename().string() << "/ - " 
              << format_size(du.size) << " (" << std::fixed << std::setprecision(2) << percentage << "%)\n";
    
    try {
        for (const auto& entry : fs::directory_iterator(path)) {
            if (entry.is_directory()) {
                print_tree_view(entry.path(), depth + 1, total_size);
            }
        }
    } catch (const fs::filesystem_error& e) {
        std::cerr << std::string((depth + 1) * 2, ' ') << "Error accessing: " << path << " (" << e.what() << ")\n";
    }
}

void print_sorted_extensions(const std::unordered_map<std::string, uintmax_t>& ext_usage) {
    std::map<uintmax_t, std::string, std::greater<>> sorted_ext;
    for (const auto& [ext, size] : ext_usage) {
        sorted_ext.insert({size, ext});
    }

    std::cout << "\nFile Extension Usage:\n";
    for (const auto& [size, ext] : sorted_ext) {
        std::cout << ext << ": " << format_size(size) << '\n';
    }
}

void print_disk_info(const fs::path& path) {
    try {
        auto space_info = fs::space(path);
        std::cout << "\nDisk Usage Information for '" << path << "':\n";
        std::cout << "Total space: " << format_size(space_info.capacity) << "\n";
        std::cout << "Used space:  " << format_size(space_info.capacity - space_info.free) << "\n";
        std::cout << "Free space:  " << format_size(space_info.free) << "\n\n";
    } catch (const fs::filesystem_error& e) {
        std::cerr << "Error retrieving disk information for: " << path << " (" << e.what() << ")\n";
    }
}

int main() {
    std::string path_str;
    std::cout << "Enter the directory path to analyze: ";
    std::getline(std::cin, path_str);
    fs::path path(path_str);

    // Print disk space info for the path
    print_disk_info(path);

    DiskUsage du = get_disk_usage(path);

    std::cout << "\nDisk Usage Tree View:\n";
    print_tree_view(path, 0, du.size);

    print_sorted_extensions(du.ext_usage);
    return 0;
}
